### ISSUE TITLE
Seeing test failures for several p4c test programs including test err/testdata/p4_16_errors/constructor3_e.p4

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4c#4822]
I run this bash script to install open source P4 dev tools on several supported versions of Ubuntu Linux monthly, and am seeing failures this month that I did not see last month.  I believe the only differences between last month and now are the versions of the behavioral-model and p4c source code, passing these tests on 2024-Jun-19 but failing on 2024-Jul-19.

+ https://github.com/jafingerhut/p4-guide/blob/master/bin/install-p4dev-v8.sh

I am seeing the same failures across all 6 combinations of processor architectures {x86_64, aarch64} and Ubuntu {20.04, 22.04, 24.04}.

Here is the output I see:
```
$ ~/p4c/build/p4test /home/ubuntu/p4c/testdata/p4_16_errors/constructor3_e.p4
p4test: /home/ubuntu/p4c/build/_deps/abseil-src/absl/container/internal/raw_hash_set.h:3154: void absl::lts_20240116::container_internal::raw_hash_set<Policy, Hash, Eq, Alloc>::emplace_at(size_t, Args&& ...) [with Args = {const IR::Type_Var*&}; Policy = absl::lts_20240116::container_internal::FlatHashSetPolicy<const IR::ITypeVar*>; Hash = Util::Hash; Eq = absl::lts_20240116::container_internal::HashEq<const IR::ITypeVar*, void>::Eq; Alloc = std::allocator<const IR::ITypeVar*>; size_t = long unsigned int]: Assertion `PolicyTraits::apply(FindElement{*this}, *iterator_at(i)) == iterator_at(i) && "constructed value does not match the lookup key"' failed.
```

It could very well be that my bash script is building open source P4 dev tools in an unsupported way, but if so, it is a way that has been working for several months now.

I believe there are a total of 26 test cases that behave similarly on all of the 6 combinations mentioned above, likely with the same root cause:
```
  2/544 Test  #822: dpdk/testdata/p4_16_dpdk_errors/pna-example-ipsec-err3.p4 ............***Failed    0.98 sec
  3/544 Test  #843: dpdk/testdata/p4_16_pna_errors/pna-example-mirror-packet-error2.p4 ...***Failed    0.89 sec
  4/544 Test  #844: dpdk/testdata/p4_16_pna_errors/pna-example-mirror-packet-error3.p4 ...***Failed    0.86 sec
  7/544 Test #2281: err/testdata/p4_16_errors/constructor3_e.p4 ..........................***Failed    0.15 sec
  8/544 Test #2306: err/testdata/p4_16_errors/factory-err2.p4 ............................***Failed    0.21 sec
  9/544 Test #2307: err/testdata/p4_16_errors/function1_e.p4 .............................***Failed    0.15 sec
 10/544 Test #2308: err/testdata/p4_16_errors/function_e.p4 ..............................***Failed    0.16 sec
 11/544 Test #2328: err/testdata/p4_16_errors/issue1006-1.p4 .............................***Failed    0.16 sec
 12/544 Test #2348: err/testdata/p4_16_errors/issue2021.p4 ...............................***Failed    0.16 sec
 13/544 Test #2350: err/testdata/p4_16_errors/issue2036-1.p4 .............................***Failed    0.15 sec
 14/544 Test #2351: err/testdata/p4_16_errors/issue2036-2.p4 .............................***Failed    0.15 sec
 15/544 Test #2352: err/testdata/p4_16_errors/issue2036.p4 ...............................***Failed    0.20 sec
 16/544 Test #2355: err/testdata/p4_16_errors/issue2230-1-bmv2.p4 ........................***Failed    0.79 sec
 17/544 Test #2357: err/testdata/p4_16_errors/issue2260-1.p4 .............................***Failed    0.16 sec
 18/544 Test #2379: err/testdata/p4_16_errors/issue3045-1.p4 .............................***Failed    0.21 sec
 19/544 Test #2380: err/testdata/p4_16_errors/issue3045-2.p4 .............................***Failed    0.20 sec
 20/544 Test #2381: err/testdata/p4_16_errors/issue3045.p4 ...............................***Failed    0.21 sec
 21/544 Test #2392: err/testdata/p4_16_errors/issue3246.p4 ...............................***Failed    0.16 sec
 22/544 Test #2414: err/testdata/p4_16_errors/issue345.p4 ................................***Failed    0.15 sec
 23/544 Test #2459: err/testdata/p4_16_errors/issue774-1.p4 ..............................***Failed    0.20 sec
 24/544 Test #2461: err/testdata/p4_16_errors/issue803-1.p4 ..............................***Failed    0.20 sec
 25/544 Test #2469: err/testdata/p4_16_errors/list-error.p4 ..............................***Failed    0.19 sec
 26/544 Test #2477: err/testdata/p4_16_errors/module_e.p4 ................................***Failed    0.16 sec
 27/544 Test #2480: err/testdata/p4_16_errors/named-fail1.p4 .............................***Failed    0.15 sec
 28/544 Test #2494: err/testdata/p4_16_errors/push_nonconstant.p4 ........................***Failed    0.19 sec
 29/544 Test #2545: err/testdata/p4_16_errors/typecheck_e.p4 .............................***Failed    0.16 sec
```
I can provide links to the full output of running the bash script linked above on these 6 systems, if anyone thinks that might be helpful in discovering what is going on.

**Comment by ChrisDodd** [comment]
These were caused by #4814 and fixed by #4818

**Comment by jafingerhut** [comment]
Great, thanks for the note.  Looks like I built just before the fix got in.   I will update and retest to verify, and close this issue if things are better as expected.

**Comment by jafingerhut** [comment]
Confirmed that with the fix in #4818 that these failures no longer occurred with a DEBUG build.  It does seem like it might be nice to experiment with adding a DEBUG build & test for _one_ supported platform choice, e.g. perhaps x86_64 Ubuntu 22.04.
